### PR TITLE
AP_Notify: Support for UAVCAN RGB LEDs v2

### PR DESCRIPTION
**Pull request by magicrub**
a rebase of https://github.com/ArduPilot/ardupilot/pull/7556

### CODE REVIEW

**Comment by magicrub** [comment]
@Programmer74 would you mind sanity checking this?

File: libraries/AP_Notify/UAVCAN_RGB_LED.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+bool UAVCAN_RGB_LED::hw_set_rgb(uint8_t red, uint8_t green, uint8_t blue)
+{
+    bool success = false;
+    if (AP_BoardConfig_CAN::get_can_num_ifaces() != 0) {
+        for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_DRIVERS; i++) {
+            if (hal.can_mgr[i] != nullptr) {
+                AP_UAVCAN *uavcan = hal.can_mgr[i]->get_UAVCAN();
+                if (uavcan != nullptr) {
+                    success |= uavcan->led_write(_led_index, red, green, blue);
```
Comment (EShamaev) [inline libraries/AP_Notify/UAVCAN_RGB_LED.cpp]: I think that catching fault should be more valuable.
Or a bitmask of successes can be used.

**Comment by EShamaev** [comment]
How the messages are enabled/disabled?
How and where the UAVCAN_RGB_LED calss is used?

**Comment by Programmer74** [comment]
@magicrub 
Seems to be working fine for my devices

**Comment by magicrub** [comment]
>How the messages are enabled/disabled?

@Programmer74  The way it looks now, we'll be sending out 20Hz LED data on everyone's CAN bus even if there's no LED hardware attached. Is that right? If so, it will need to change to only send if manually enabled and/or device is detected/attached.

**Comment by Programmer74** [comment]
@magicrub 

Not on every device's bus, but to those that have AP_UAVCAN_RGBLED notification device enabled
The only device with this notification will be Edge.

**Comment by magicrub** [comment]
@Programmer74 which none are. Did you mean to add UAVCAN_RGBLED [in AP_Notify](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Notify/AP_Notify.cpp#L168) for the LINUX_EDGE device?

**Comment by Programmer74** [comment]
@magicrub 
Yes, but not in this PR.
We were planning another PR with enabling UAVCAN and other our UAVCAN features for LINUX_EDGE when everything gets merged.

**Comment by magicrub** [comment]
@Programmer74 Thanks for checking the branch. @EShamaev looks ready to merge

**Comment by EShamaev** [comment]
I would prefer to see the possibility to enable/disable it before merging and integration into AP_Notify.
However as I see from code if the led_write is not called - no enable is set so should be fine, but without integration into AP_Notify it seems to be useless.
Also I still have concerns about return value of hw_set_rgb.

File: libraries/AP_Notify/UAVCAN_RGB_LED.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_Notify/UAVCAN_RGB_LED.cpp]: I agree, but that's not worth holding up the PR. We can add that later.. maybe to all if the modules

**Comment by EShamaev** [comment]
My point is that the PR should add real functionality and be complete. Right now it adds mechanism but it is not used anywhere.
Also this feature will be added to all boards, so support in AP_Notify should not be only for "your Edge board" or only Linux devices.

As for PR I have no objections to code, but I think it is incomplete.

**Comment by magicrub** [comment]
@programmer74 would you like me to add the one line in ap_notify.cpp to enable this for the Linux edge device? Otherwise this is dead code.

I'm just getting on a plane to Australia. You can grab my branch and do it yourself in a V3 PR or else I'll add it after my loooong flight.

**Comment by magicrub** [comment]
I tried adding the driver to AP_Notify.cpp but didn't know what to use for the index. Maybe iterated it for all indexes? Not sure.

`    ADD_BACKEND(new UAVCAN_RGB_LED(?));
`

**Comment by Programmer74** [comment]
@EShamaev @magicrub The Edge even has no UAVCAN enabled by now, so it shouldn't be added to AP_Notify right now. Adding UAVCAN in general and this notification device in particular for Edge right now in this PR will break logic of "one PR = one thing involved"
This won't be the "dead" code for a long time.

**Comment by magicrub** [comment]
@eshanaev I'm in favor of merging. It will get used soon. UAVCAN has a lot of work happening on it, I'm confident it will get enabled correctly soon.

**Comment by EShamaev** [comment]
OK, no problem, but I do not understand why this PR is discussed only with consideration of Linux and Edge?
If it is really so, wrap the code in UAVCAN lib in defines that you will enable in later PRs so that you do not clatter other builds with dead code.

**Comment by Programmer74** [comment]
@EShamaev Of course, it can be used with non-Linux boards, we just have the implementation of the device supporting it almost ready.

**Comment by EShamaev** [comment]
I see that contradiction:
_"Enabling this as a notification device for edge has no sense until that PR will be merged"
"Of course, it can be used with non-Linux boards"_

Then my position is that the PR should be complete in all terms and add full support of the UAVCAN RGB LED that can be used dependless on board and other PRs.

**Comment by Programmer74** [comment]
This can be used and merged dependless on board and other PRs.

I say that this code will be "dead" until there is any device that supports it. The first device that should have this support is Edge, but until the mentioned PR gets merged this PR won't work on Edge.

**Comment by EShamaev** [comment]
Until the AP_Notify is modified to support it, it is dead.
@magicrub I will leave the decision up to you, but in my opinion, without modification of AP_Notify, this is useless and I do not see any objections from including the modification of AP_Notify in this PR to support this device.

**Comment by Programmer74** [comment]
> and I do not see any objections from including the modification of AP_Notify in this PR to support this device

In the current master branch, Edge has no CAN support at all. It won't work even if we modify AP_Notify for Edge.
Modifying AP_Notify for other boards... I don't think there is any sense in this PR either.

We could include ```UAVCAN_RGB_LED.h``` in AP_Notify, but without ```ADD_BACKEND``` for any board, though.

**Comment by magicrub** [comment]
Ok. So. The current status is this PR is ready to be made as long as it's not dead code. However, even your other PR that adds Edge support doesn't add support for this.. Becka's it doesn't exist yet. Chicken egg problem.

Can you make a single PR that adds everything to make the edge work which makes everyone happy?

**Comment by Programmer74** [comment]
@magicrub Waiting for that other PR to make use if it here is the best that we can do now.

**Comment by magicrub** [comment]
@programmer74 the requested changes never happened on the other PR. If you need help, I can do them for you to get it done fast.

**Comment by Programmer74** [comment]
@magicrub Pick the commit https://github.com/Programmer74/ardupilot/commit/97b8c922914b929f75d5f5e1e4cd95c7fb44d78f to this PR and this will be ready to be merged.

**Comment by magicrub** [comment]
@Programmer74  done. @EShamaev That should be ready to merge now.

**Comment by magicrub** [comment]
SemaphoreCI failing is normal and can (temporarily) be ignored.

**Comment by Linjieqiang** [comment]
Where can I get the UAVCAN RGB module to test?

**Comment by Linjieqiang** [comment]
Where can I get the UAVCAN RGB module to test?

**Comment by Programmer74** [comment]
@Linjieqiang For now, only Edge GNSS module with an updated firmware can act as such module.

**Comment by Linjieqiang** [comment]
@Programmer74 I want to make a UAVCAN module with RGB LED and Beeper.Has anyone interested in the firmware coding for it?Please contact me.

**Comment by magicrub** [comment]
@EShamaev please see commit https://github.com/ArduPilot/ardupilot/pull/7680/commits/0ee1f5c5963543386f4552b3a016730f1383ff2c which enables this. Should be ready to merge.

**Comment by EShamaev** [comment]
Should not have add_backend been wrapped in ifdef's?
Anyway, seems good now and ready to be used.

**Comment by Programmer74** [comment]
@magicrub ping?

**Comment by magicrub** [comment]
oh geez, sorry. Been traveling a bunch and lost track of this. I'll rebase and merge.

**Comment by magicrub** [comment]
merged!

**Comment by skyyuzhang** [comment]
When I enable my uavcan . I can not get any data about RGB LED but Nodestatus  on UAVCAN GUI. also  it is not work when I set NTF_LED_TYPES:LED to 5 (UAVCAN) is there an setting I missed? 
CAN_D1_Protocol =1
CAN_D1_UC_NODE=10
CAN_D1_UC_ESC_BM=15
CAN_P1_BITRATE= 1000000

**Pull request merged**